Feature/osw auto addup spec

singjc opened the pull request:
## Description

This is 1 of 3 (probably) PRs related to improvements for ion mobility IPF scoring. I'm breaking up my main implementation into smaller PRs to hopefully make the revision easier.

This PR implements the ability to automatically add up spectra across the peak width in retention time. The original implementation extracted spectra at the peak apex, with the option of extracting more spectra across the peak by setting the `add_up_spectra_` param to an odd value greater than 1. Now the user can set `add_up_spectra_` to -1, which will turn on the automatic computation of `add_up_spectra_` based on the percent of the peak width. The general idea behind adding up more spectra, is that it could lead to cleaner / higher signal in the ion mobility extraction.

![image](https://github.com/user-attachments/assets/3ec34e94-8949-4390-974e-94f650a20890)

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

singjc committed:
add: getMetaValue method to IMRMFeature interface

singjc committed:
add: getMetaValue method to MRMFeatureOpenMS

singjc committed:
update,add: add_up_spectra arg doc and add use_percent_peak_width arg/var for automatrically determining add_up_spectra based on RT peak width

singjc committed:
chore: remove sonar swath map check for calculateDIAIdScores. See PR #7722

singjc committed:
add: auto add up spectra logic to DIA ipf scoring

timosachsenberg commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/OpenSwathScoring.h:
```diff
@@ -71,6 +72,7 @@ namespace OpenMS
      * @param rt_normalization_factor Specifies the range of the normalized retention time space
      * @param add_up_spectra How many spectra to add up (default 1)
      * @param spacing_for_spectra_resampling Spacing factor for spectra addition
+     * @param use_percent_peak_width Percent to constrict the number of spectra to add based on peak width
```
```suggestion
     * @param use_percent_peak_width Percent to construct the number of spectra to add based on peak width
```

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFinderScoring.cpp:
```diff
@@ -72,10 +72,13 @@ namespace OpenMS
     defaults_.setValidStrings("spectrum_addition_method", {"simple", "resample"});
-    defaults_.setValue("add_up_spectra", 1, "Add up spectra around the peak apex (needs to be a non-even integer)", {"advanced"});
-    defaults_.setMinInt("add_up_spectra", 1);
+    defaults_.setValue("add_up_spectra", 1, "Add up spectra around the peak apex (needs to be a non-even integer). If set to -1, it will automatically select the number of spectra to add up around the peak apex.", {"advanced"});
```
```suggestion
    defaults_.setValue("add_up_spectra", 1, "Add up spectra around the peak apex (needs to be a non-even integer). If set to -1, it will automatically select the number of spectra to add up around the peak apex according to use_percent_peak_width.", {"advanced"});
```

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFinderScoring.cpp:
```diff
@@ -72,10 +72,13 @@ namespace OpenMS
+    defaults_.setMinInt("add_up_spectra", -1);
     defaults_.setValue("spacing_for_spectra_resampling", 0.005, "If spectra are to be added, use this spacing to add them up", {"advanced"});
     defaults_.setMinFloat("spacing_for_spectra_resampling", 0.0);
+    defaults_.setValue("use_percent_peak_width", 0.15, "If spectra are to be added based on the peak width of peak, constrict number of spectra to be added based on N percent of number of points of peak width.", {"advanced"});
```
should we call it percent if it is a fraction?

singjc replied:
Good point, could probably change it to `fraction_of_peak_width`?

timosachsenberg replied:
intend could be clearer if we use something like "merge_spectra_by_peak_width_fraction"

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -86,6 +89,33 @@ namespace OpenMS
+        add_up_spectra_ = std::ceil(peakWidth * use_percent_peak_width_);
+        
+        // Ensure add_up_spectra_ is odd
+        if (static_cast<int>(add_up_spectra_) % 2 == 0) {
```
I would change the value such that we don't need to have these kind of checks.
e.g. if  +-add_up_spectra are extracted, we would not need such a check and "hack" to make it a valid ordinal range to extract

singjc replied:
I'm not sure if this is what you meant, or the best way to do this, but I now do:

```
n_merge_spectra = std::max(1, static_cast<int>(std::ceil(peakWidth * merge_spectra_by_peak_width_fraction_)) | 1);
```

- Use 1 if the resulting expression on the right is 0 or less
- if the calculation of ceiling peak width * N frac is even, use bitwise OR operation to ensure it's odd.

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -86,6 +89,33 @@ namespace OpenMS
+            add_up_spectra_ -= 1;
+        }
+        
+        // Ensure add_up_spectra_ is at least 1 if it was set to <= 0 based on the peak width
```
I think this is dangerous. Changing the value that represents something stored in the internal param object can easily lead to unexpected and hard to debug bugs. Imagine (this arificial example) that e.g., another thread would update a totally unrelated parameter of the Param object. This would call updateParameters of that class (which might set add_up_spectra_ to 0 again).
It would be really hard to find bugs related to add_up_spectra_ suddenly being 0.
Also, I think values other than -1 don't have a semantic meaning here. So the correct way would be to fail for all non-semantically valid values.

timosachsenberg replied:
so it can never be -1? after this line?

singjc replied:
This is a good point. I think your comment below for setting a local variable would be better than modifying the param object. 

The actual value for add_up_spectra needs to be a positive value, otherwise `fetchSpectrumSwath` will fail. I use -1 to identify whether `add_up_spectra` should be computed dynamically, but maybe that would cause some downstream errors?

singjc replied:
Should I add another arg to to turn on the auto `add_up_spectra`? Or change `add_up_spectra` to be a string option?

timosachsenberg replied:
I would probably not change it into a string.
e.g. one could define:
```
enum class SpectrumMergeMethodType { Fixed, Dynamic };
and change the initialize function to take both the SpectrumMergeMethodType and a double merge_window_size
```
and then query this information in the calculateDIAScores method.
Depending on the SpectrumMergeMethodType this will fill the local variable e.g. n_merge_spectra with a constant value or dynamic by peak width.

singjc replied:
> I would probably not change it into a string. e.g. one could define:
> 
> ```
> enum class SpectrumMergeMethodType { Fixed, Dynamic };
> and change the initialize function to take both the SpectrumMergeMethodType and a double merge_window_size
> ```
> 
> and then query this information in the calculateDIAScores method. Depending on the SpectrumMergeMethodType this will fill the local variable e.g. n_merge_spectra with a constant value or dynamic by peak width.

Okay, I have implemented something along these lines now.

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -252,27 +282,35 @@ namespace OpenMS
-    std::vector<OpenSwath::SwathMap> used_swath_maps;
-    if (swath_maps.size() > 1)
+    // If add_up_spectra_ is -1, we automatically compute the amount of spectra to add by setting add_up_spectra_ to the ceiling value of N percent of the retention time peak width
+    if (add_up_spectra_ == -1)
```
see that this is highly dangerous.
the method above changes add_up_spectra_ to a value != -1 if called before this method

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -252,27 +282,35 @@ namespace OpenMS
-      used_swath_maps = swath_maps;
+        
+        // Ensure add_up_spectra_ is at least 1 if it was set to <= 0 based on the peak width
+        if (add_up_spectra_ <= 0) {
```
same issue here

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -252,27 +282,35 @@ namespace OpenMS
 
     // find spectrum that is closest to the apex of the peak using binary search
-    std::vector<OpenSwath::SpectrumPtr> spectrum = fetchSpectrumSwath(used_swath_maps, imrmfeature->getRT(), add_up_spectra_, im_range);
+    std::vector<OpenSwath::SpectrumPtr> spectrum = fetchSpectrumSwath(swath_maps, imrmfeature->getRT(), add_up_spectra_, im_range);
```
would be better to pass a local variable instead of a modified member.
the member variable should represent a property of the scoring object but add_up_spectra is now a property that depends on the individual feature.

timosachsenberg replied:
so maybe a local variable spectra_merge_window that gets set based on the add_up_spectra_ / percentage settings

timosachsenberg commented:
nice addition. I left some comments and added other reviewers.

singjc committed:
Update src/openms/include/OpenMS/ANALYSIS/OPENSWATH/OpenSwathScoring.h

Co-authored-by: Timo Sachsenberg <timo.sachsenberg@uni-tuebingen.de>

singjc committed:
Update src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFinderScoring.cpp

Co-authored-by: Timo Sachsenberg <timo.sachsenberg@uni-tuebingen.de>

singjc committed:
update: use_percent_peak_width to merge_spectra_by_peak_width_fraction_

singjc committed:
add: SpectrumMergeMethodType enum to specify fixed or dynamic number of spectra to merge

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFinderScoring.cpp:
```diff
@@ -72,10 +72,15 @@ namespace OpenMS
-    defaults_.setValue("add_up_spectra", 1, "Add up spectra around the peak apex (needs to be a non-even integer)", {"advanced"});
+    defaults_.setValue("spectrum_merge_method_type", "fixed", "For spectrum addition, either use a fixed number of spectra or dynamically select the number of spectra to add around the peak apex based on the merge_spectra_by_peak_width_fraction.", {"advanced"});
+    defaults_.setValidStrings("spectrum_merge_method_type", {"fixed", "dynamic"});
+    defaults_.setValue("add_up_spectra", 1, "Add up spectra around the peak apex (needs to be a non-even integer).", {"advanced"});
```
I was referring to this uneven number as requirement. If you change this plus and minus you don’t need the % 2 check.

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -86,8 +104,28 @@ namespace OpenMS
+        double leftWidth = imrmfeature->getMetaValue("leftWidth");
+        double rightWidth = imrmfeature->getMetaValue("rightWidth");
+        double peakWidth = rightWidth - leftWidth;
+        n_merge_spectra = std::max(1, static_cast<int>(std::ceil(peakWidth * merge_spectra_by_peak_width_fraction_)) | 1);
```
Same here. You would get rid of the hacky | 1

timosachsenberg replied:
ah ok I see that you now have separate arguments in the initialize method.
I think it might make sense to have a quick zoom call at some point so I can better explain how this could be approached. I am terrible at explaining this as part of a GitHub issue :)

singjc replied:
Sounds good. I just made another PR for IM peak picking, I have to make one more PR for the rest of the IM-IPF improvements. Maybe we can have a zoom call after I've made the 3rd PR, so if there are other suggestions/fixes for those, then we can discuss those at the same time as well? Maybe this Wednesday or Thursday?

timosachsenberg replied:
Works for me. I am Meeting with Josh on Thursday anyway so maybe we Can scheffle around that time?

singjc replied:
> Works for me. I am Meeting with Josh on Thursday anyway so maybe we Can scheffle around that time?

Sounds good. @jcharkow  could we use the same zoom room?

jcharkow replied:
Yes we can use the same zoom room. Room+Time is in the lab calendar.

jcharkow reviewed (approved):
Overall looks good to me. 

One thought though is does it make sense to have both the "fixed" and the "percentile" based addition approaches? Would it make more sense to replace the fixed with the percentile approach or are there cases where the fixed approach is better?

I do not think the fixed approach is used in practice, at least for me.

timosachsenberg commented:
> Overall looks good to me.
> 
> One thought though is does it make sense to have both the "fixed" and the "percentile" based addition approaches? Would it make more sense to replace the fixed with the percentile approach or are there cases where the fixed approach is better?
> 
> I do not think the fixed approach is used in practice, at least for me.

I agree that this would also be my preferred way forward if possible. E.g., only support percentage
with a default parameters (e.g., something along FWHM) and just have reasonable internal boundaries (e.g. a minimum of 1 and a maximum of 5 scans - something that significantly reduces noise but does not increase computation too much).

singjc commented:
> Overall looks good to me.
> 
> One thought though is does it make sense to have both the "fixed" and the "percentile" based addition approaches? Would it make more sense to replace the fixed with the percentile approach or are there cases where the fixed approach is better?
> 
> I do not think the fixed approach is used in practice, at least for me.

I do agree as well, though I've left the fixed as an option because that's the current approach that is used. But we could make the dynamic merging of spectra the default? Or like Timo suggested, we could have a default where we just clamp between a reasonable range. This would probably be important for cases where we have a really large peak.

timosachsenberg commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/DATAACCESS/MRMFeatureAccessOpenMS.h:
```diff
@@ -75,6 +75,8 @@ namespace OpenMS
 
     double getRT() const override;
 
+    double getMetaValue(std::string name) const;
```
consider using RangeRT

singjc replied:
I was looking into applying this, but I think it would require more changes to other parts of OpenSwath. Currently OpenSwath (and some of the other FeatureFinders) stores the leftWidth and rightWidth in the metaValue, which is then used for retrieval at some point. I think the `RangeRT` would be a good replacement, but I think those changes would maybe be for another PR?

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp:
```diff
@@ -252,27 +290,28 @@ namespace OpenMS
 
     // find spectrum that is closest to the apex of the peak using binary search
-    std::vector<OpenSwath::SpectrumPtr> spectrum = fetchSpectrumSwath(used_swath_maps, imrmfeature->getRT(), add_up_spectra_, im_range);
+    std::vector<OpenSwath::SpectrumPtr> spectrum = fetchSpectrumSwath(swath_maps, imrmfeature->getRT(), n_merge_spectra, im_range);
```
could probably be  one

singjc committed:
chore,doc: make add up spec doc more clearer, to add up spectra by equal amounts on both sides.

singjc committed:
fix,clean: remove hack to enfore odd number for n_merge_spectra

timosachsenberg committed:
Merge branch 'develop' into feature/osw_auto_addup_spec

timosachsenberg commented:
> > Overall looks good to me.
> > One thought though is does it make sense to have both the "fixed" and the "percentile" based addition approaches? Would it make more sense to replace the fixed with the percentile approach or are there cases where the fixed approach is better?
> > I do not think the fixed approach is used in practice, at least for me.
> 
> I do agree as well, though I've left the fixed as an option because that's the current approach that is used. But we could make the dynamic merging of spectra the default? Or like Timo suggested, we could have a default where we just clamp between a reasonable range. This would probably be important for cases where we have a really large peak.

Yeah, less options = better test coverage, less combinatorics => less things that can get wrong

timosachsenberg commented:
I think the mock mrmfeature also needs the new method implemented

singjc committed:
Merge branch 'develop' into feature/osw_auto_addup_spec

singjc commented:
Sorry, been busy with another project. Just wanted to follow up on this, I think I mostly addressed all the comments/suggestions? You mentioned the mock MRMFeature needed the new method implementation? Can you point me to where that is?

timosachsenberg commented:
This is the error: ../src/tests/class_tests/openms/source/MRMScoring_test.cpp:590:46: error: allocating an object of abstract class type 'OpenSwath::MockMRMFeature'
https://github.com/OpenMS/OpenMS/blob/864b4e52c5e7b86c07d627e072f6d5f24063f835/src/openswathalgo/include/OpenMS/OPENSWATHALGO/DATAACCESS/MockObjects.h#L51
Probably an added method is not implemented there.

singjc committed:
add,fix: getMetaValue to MockObjects

singjc committed:
Merge branch 'feature/osw_auto_addup_spec' of github.com:singjc/OpenMS into feature/osw_auto_addup_spec

singjc commented:
> This is the error: ../src/tests/class_tests/openms/source/MRMScoring_test.cpp:590:46: error: allocating an object of abstract class type 'OpenSwath::MockMRMFeature'
> 
> https://github.com/OpenMS/OpenMS/blob/864b4e52c5e7b86c07d627e072f6d5f24063f835/src/openswathalgo/include/OpenMS/OPENSWATHALGO/DATAACCESS/MockObjects.h#L51
> 
> Probably an added method is not implemented there.

Ahh kk, I updated the MockObjects. Should work now, passed locally.

singjc committed:
fix,update: OpenSwathScoring test, with updated params in intializer

timosachsenberg commented on src/openswathalgo/include/OpenMS/OPENSWATHALGO/DATAACCESS/ITransition.h:
I think these interfaces would profit a lot from some documentation in the future.
e.g. such that things like the difference between:
    virtual void getIntensity(std::vector<double>& intens) const = 0;
    virtual float getIntensity() const = 0;

for what virtual double getMetaValue(std::string name) const = 0; is used (and ideally which ones are currently queried)

singjc committed:
Merge branch 'develop' into feature/osw_auto_addup_spec

singjc commented:
All tests pass, can we merge?
